1/x Encourage admin users to connect to a database

WiNloSt opened the pull request:
Epic: https://github.com/metabase/metabase/issues/31015
[Product doc](https://www.notion.so/metabase/Really-encourage-admins-to-add-a-DB-9c8c3a7d08ca45cb8672a154220f443e?pvs=4)

### Description

This PR adds a banner on top of the app bar when these conditions are met
1. User is an admin
2. A Metabase instance is on a paid plan
3. The instance only has a sample database.

There is another variation of the banner that we will not show `Connect your database` button when users is on `/admin/databases/create` otherwise, this button should be visible at all time (given all 3 conditions are met still).

### How to verify

Try to play around with the combinations of the above conditions, only when the 3 conditions are met then the banner would show up.

### Demo
#### App
![image](https://github.com/metabase/metabase/assets/1937582/d833571c-e218-4eb2-8892-e51d5622e1b0)

#### Admin add database page
![image](https://github.com/metabase/metabase/assets/1937582/09ee2653-c24e-48f8-b287-238555528035)

#### Other admin pages
![image](https://github.com/metabase/metabase/assets/1937582/a0eaba28-ec96-4d64-93a7-7807493a3ad8)


### Checklist

- [x] Tests have been added/updated to cover changes in this PR

WiNloSt committed:
Encourage admin users to connect to a database

kdoh reviewed (approved):
Looking good to me. 👍

WiNloSt committed:
Only render either payment banner or database prompt banner

Discussed here:
https://metaboat.slack.com/archives/C057T1QTB3L/p1685549180307149?thread_ts=1685547913.580079&cid=C057T1QTB3L

WiNloSt committed:
Improve mobile view layout

WiNloSt commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.styled.ts:
```diff
@@ -0,0 +1,60 @@
+  align-items: center;
+  flex-wrap: wrap;
+
+  ${breakpointMinSmall} {
```
Add breakpoints to various elements to make the mobile view looks more polished. [The design is here](https://www.figma.com/file/H8NL31kXy2bNxuXtSKX0Fu/Really-encourage-admins-to-add-a-DB?type=design&node-id=27-92&t=ktyrZcQNXHgbNUrR-0).

#### Before
- ![image](https://github.com/metabase/metabase/assets/1937582/ebf5f156-088b-4e88-aefa-c2d0bb1c6145)
- ![image](https://github.com/metabase/metabase/assets/1937582/10ae7ab5-231a-459a-8019-fe54fde12790)

#### After
- ![image](https://github.com/metabase/metabase/assets/1937582/aa56a297-aee7-4159-98f8-990b5896a77f)
- ![image](https://github.com/metabase/metabase/assets/1937582/ce2adf87-676f-4b0a-ba93-4e0b5ed36de7)

WiNloSt commented on frontend/src/metabase/components/AppBanner/AppBanner.tsx:
```diff
@@ -0,0 +1,19 @@
+  const tokenStatusStatus = useSelector(
+    state => getSetting(state, "token-status")?.status,
+  );
+  if (PaymentBanner.shouldRender({ isAdmin, tokenStatusStatus })) {
```
Only render one banner at a time, and if somehow both banner conditions are met, only render the payment banner. [Discussion here.](https://metaboat.slack.com/archives/C057T1QTB3L/p1685549180307149?thread_ts=1685547913.580079&cid=C057T1QTB3L)

ranquild commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.tsx:
```diff
@@ -0,0 +1,59 @@
+  location: Location;
+}
+
+export const DatabasePromptBanner = withRouter(
```
HoCs are discouraged, could you pass `Location` from the nearest component rendered by the router instead?

ranquild commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.tsx:
```diff
@@ -0,0 +1,59 @@
+  function DatabasePromptBannerView({ location }: DatabasePromptBannerProps) {
+    const isAdmin = useSelector(getUserIsAdmin);
+    const isPaidPlan = useSelector(getIsPaidPlan);
+    const { data: databases } = useDatabaseListQuery({
```
nit `data: databases = []` if you want to default an empty list

ranquild commented on frontend/src/metabase/nav/components/PaymentBanner/PaymentBanner.tsx:
```diff
@@ -0,0 +1,54 @@
+  return null;
+};
+
+PaymentBanner.shouldRender = ({
```
Attaching functions to components isn't something common on the frontend, could you please make this a regular function instead?

ranquild reviewed (commented):
@WiNloSt could you add one E2E test that verifies you can see the new banner, click on the button and see the form to add a database? Otherwise the PR looks good!

WiNloSt committed:
Address review: Remove `withRouter` HoC

WiNloSt committed:
Address review: Remove optional chaining

WiNloSt committed:
Address review: Refactor a component's deprecated `shouldRender` property

WiNloSt committed:
Add an e2e test to cover happy path

WiNloSt commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <DatabasePromptBannerRoot
+      role="banner"
+      aria-label={t`Database prompt banner`}
```
@mazameli @ranquild I'm not sure this warrant a new translation. The label is only used for the E2E test, (and hopefully for screen readers too).

Please feel free to suggest a better name if it should be translated.

ranquild replied:
I wouldn't add an aria-label here, at least not normally. We can either add `data-testid` or just check for something in the banner.

WiNloSt replied:
Removed, and use `cy.findAllByRole("banner").first()` to target this banner instead.

WiNloSt commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.tsx:
```diff
@@ -0,0 +1,59 @@
+import { t } from "ttag";
```
@ranquild I added one test that test a happy path of this feature, the rest should already be covered with the integration test.

WiNloSt committed:
Remove new translation in `aria-label`

WiNloSt committed:
Fix sticky parameter widgets when showing database prompt banner

WiNloSt committed:
Fix E2E tests

WiNloSt commented on frontend/src/metabase/components/AppBanner/AppBanner.jsx:
The file changed to TypeScript. And move the content to `PaymentBanner` instead.

WiNloSt commented on frontend/src/metabase/components/AppBanner/AppBanner.tsx:
```diff
@@ -0,0 +1,28 @@
+  const tokenStatusStatus = useSelector(
+    state => getSetting(state, "token-status")?.status,
+  );
+  if (shouldRenderPaymentBanner({ isAdmin, tokenStatusStatus })) {
```
Only render one banner at a time, and if somehow both banner conditions are met, only render the payment banner. [Discussion here.](https://metaboat.slack.com/archives/C057T1QTB3L/p1685549180307149?thread_ts=1685547913.580079&cid=C057T1QTB3L)

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.jsx:
```diff
@@ -332,7 +331,9 @@ class Dashboard extends Component {
               <ParametersAndCardsContainer
                 data-testid="dashboard-parameters-and-cards"
-                ref={element => (this.parametersAndCardsContainerRef = element)}
+                shouldMakeDashboardHeaderStickyAfterScrolling={
```
This prop controls whether we will render this container with `overflow: auto` or not
<img width="1453" alt="image" src="https://github.com/metabase/metabase/assets/1937582/19e8e9bd-4937-4d32-b740-b61d0e01519a">

If we want the dashboard header to be sticky as well. We need to set `overflow-y: auto`

Here are some variations of the UI

1. Long dashboard with more info panel open
<img width="1451" alt="image" src="https://github.com/metabase/metabase/assets/1937582/066c1d58-ef02-485a-9878-a37bedf28086">
2. Long dashboard with subscription sidebar open
<img width="1453" alt="image" src="https://github.com/metabase/metabase/assets/1937582/4cab19e6-35f0-487c-9c92-206e9e768c1b">
3. Editing dashboard
<img width="1451" alt="image" src="https://github.com/metabase/metabase/assets/1937582/9fda2ee2-5a29-4bac-a6bc-e73ab95a61e2">

4. Full-screen dashboards
<img width="1512" alt="Screenshot 2023-06-07 at 3 09 48 PM" src="https://github.com/metabase/metabase/assets/1937582/19760e30-3200-4858-9da5-6e9f8c9dd797">


This follows the current behaviors on the dashboard, I think overall the whole UI should be standardized since you could probably see that we scroll on different containers and make different UIs sticky as well.

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.jsx:
```diff
@@ -87,7 +80,12 @@ export const HeaderContainer = styled.header`
+    shouldMakeDashboardHeaderStickyAfterScrolling ? "auto" : "visible"};
   overflow-x: hidden;
+  @supports (overflow-x: clip) {
+    overflow-x: clip;
```
For the rest [10% of browsers](https://caniuse.com/?search=overflow%3Aclip) that don't support `overflow-x: clip`, we just don't make the parameter sticky at all (please note that in different layouts where we don't hide the dashboard header, the parameters will still be sticky together with the dashboard header). [Bruno is okay with this](https://metaboat.slack.com/archives/C505ZNNH4/p1686054411768389?thread_ts=1686054215.543739&cid=C505ZNNH4).

#### 90% of browsers
<img width="1451" alt="image" src="https://github.com/metabase/metabase/assets/1937582/d4c833db-43f4-453b-8178-a8e969e9401a">

#### 10% of browsers
<img width="1450" alt="image" src="https://github.com/metabase/metabase/assets/1937582/37586b34-4547-44a9-9775-d39326e836fa">

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/stickyParameters.js:
```diff
@@ -13,8 +13,6 @@ export const updateParametersWidgetStickiness = dashboard => {
   );
 
   if (shouldToggleStickiness) {
-    updateParametersAndCardsContainerStyle(dashboard, shouldBeSticky);
```
This isn't needed because it's handled by `position: sticky`

WiNloSt commented on frontend/src/metabase/dashboard/components/Dashboard/stickyParameters.unit.spec.js:
```diff
@@ -15,7 +15,6 @@ describe("updateParametersWidgetStickiness", () => {
 
     const dashboard = {
       parametersWidgetRef: { offsetTop },
-      parametersAndCardsContainerRef: { style: {} },
```
update tests since we don't use this anymore. This was used when we weren't using `position: sticky` and rely on JS positional calculations.

WiNloSt commented on frontend/src/metabase/nav/components/PaymentBanner/PaymentBanner.tsx:
```diff
@@ -0,0 +1,54 @@
+import { jt, t } from "ttag";
```
This should look mostly like the previous version of the `AppBanner`
https://github.com/metabase/metabase/blob/fdc07959b9369547221ac5ab897a951f24160f46/frontend/src/metabase/components/AppBanner/AppBanner.jsx

The only notable change is the logic to determine whether we should display this banner is lifted up, so we can decide when to display this banner or the database prompt banner.

ranquild commented on e2e/test/scenarios/dashboard/dashboard.cy.spec.js:
```diff
@@ -693,7 +693,9 @@ function checkOptionsForFilter(filter) {
-  cy.findByText("Pick one or more filters to update").click();
+  // Uses force: true because the popover is covering this text. This happens
+  // after we introduce the database prompt banner.
+  cy.findByText("Pick one or more filters to update").click({ force: true });
```
Can we scroll to it somehow? e.g. scrollIntoView or similar

WiNloSt replied:
This is the place that this test is clicking
![Screenshot 2023-06-08 at 4 36 09 PM](https://github.com/metabase/metabase/assets/1937582/e31e3dd2-0def-4e9f-bd14-852200ac0f2f)

Unfortunately, this is a full-height layout. So it's impossible to scroll anywhere on this page for this particular setup. The problem is when clicking the **Category** filter. It will look like the following screenshot which completely cover the text element we want to click.
![image](https://github.com/metabase/metabase/assets/1937582/d9123979-a485-4725-a5bc-a3bf84295bf4)

ranquild commented on frontend/src/metabase/dashboard/components/Dashboard/Dashboard.styled.jsx:
```diff
@@ -87,7 +80,12 @@ export const HeaderContainer = styled.header`
 
 export const ParametersAndCardsContainer = styled.div`
   flex: auto;
+  overflow-y: ${({ shouldMakeDashboardHeaderStickyAfterScrolling }) =>
```
nit: shorter parameter name? :)

WiNloSt replied:
The other approach in this file is when we specify what we pass instead. e.g. `isEditing or `isEditingOrSharing` which won't work with what we're passing here. `!isFullscreen && (isEditing || isSharing)` so I think it's better to document what we want this flag to mean.

I'm open to new names. But this is the most concise version I could come up with at the time. I think the weirdness with the name would be getting away once we start making dashboard UI header + sticky components more consistent. You can see [in my other comment](https://github.com/metabase/metabase/pull/31210#discussion_r1221121046) that we have quite different variations for dashboard headers. Notably different elements will be sticky depending on whether we're opening a subscription sidebar or the info sidebar.

ranquild commented on frontend/src/metabase/components/AppBanner/AppBanner.tsx:
```diff
@@ -0,0 +1,28 @@
+  location: Location;
+}
+
+export const AppBanner = ({ location }: AppBannerProps) => {
```
I don't see tests for this component. Do they exist?

WiNloSt replied:
They don't exist. This component used to house what's in the `PaymentBanner` which is newly introduced in this PR. The only tests for the previously `AppBanner` aka `PaymentBanner` (now) are [E2E tests](https://github.com/metabase/metabase/blob/37a24a471908b2caf64698418614a905a80e4d66/e2e/test/scenarios/admin/subscription/payment-failure.cy.spec.js#L11).

Would it make sense to add a few assertions in that file `e2e/test/scenarios/admin/subscription/payment-failure.cy.spec.js` to assert that the database prompt banner won't be displayed, because the database prompt banner will only display when the payment banner isn't showing?

WiNloSt replied:
Added tests

ranquild commented on frontend/src/metabase/nav/components/PaymentBanner/PaymentBanner.tsx:
```diff
@@ -0,0 +1,54 @@
+  tokenStatusStatus: string | undefined;
+}
+
+export const PaymentBanner = ({
```
I don't see tests for this component. Do they exist?

WiNloSt replied:
They exist. Here are the tests https://github.com/metabase/metabase/blob/master/e2e/test/scenarios/admin/subscription/payment-failure.cy.spec.js.

[Here's another related comment](https://github.com/metabase/metabase/pull/31210#discussion_r1222738533), since this component used to be named `AppBanner`

WiNloSt replied:
Added tests

ranquild commented on frontend/src/metabase/selectors/settings.ts:
```diff
@@ -40,3 +40,12 @@ const getUtmSource = (features: TokenFeatures) => {
+
+const getTokenFeatures = (state: State) => getSetting(state, "token-features");
+
+export const getIsPaidPlan = createSelector(
```
I don't see tests for this utility. Do they exist?

WiNloSt replied:
Good catch. They don't, since I just created this selector in this PR. I will add one.

WiNloSt replied:
Added tests

ranquild reviewed (commented):
Looks good, comments about testing only

WiNloSt committed:
Address review: Add missing unit tests

WiNloSt commented:
@ranquild I added 3 more unit test files. please see this commit https://github.com/metabase/metabase/pull/31210/commits/19cc26b0be3c11a28d2fae37163d41a7662afff2

Those tests should cover every case possible.

WiNloSt committed:
Fix a failed E2E test from rebasing off `master` branch

WiNloSt commented on e2e/test/scenarios/admin/datamodel/editor.cy.spec.js:
```diff
@@ -633,7 +633,7 @@ const getFieldSection = fieldName => {
 
 const moveField = (fieldIndex, deltaY) => {
-  cy.get(".Grabber").eq(fieldIndex).trigger("mousedown", 0, 0);
+  cy.get(".Grabber").eq(fieldIndex).trigger("mousedown", 0, 0, { force: true });
```
Interestingly,[ this just fails](https://www.deploysentinel.com/ci/runs/6481ca4014a47e6dd93d5a65) on the latest commit https://github.com/metabase/metabase/pull/31210/commits/19cc26b0be3c11a28d2fae37163d41a7662afff2, which only added unit tests 🤔 . Maybe it has something to do with rebasing off `master` too.

But I think this is justified since other place that `trigger("mousedown", xxx)` all uses `force: true`
![Screenshot 2023-06-08 at 9 24 43 PM](https://github.com/metabase/metabase/assets/1937582/e328f8ca-92b7-44ff-b1ca-e34d6a9431e4)

ranquild commented on frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:
```diff
@@ -0,0 +1,212 @@
+      renderingBanner: "Unpaid banner",
+    },
+  ] as const)(
+    "should render $renderingBanner when should render past-due banner: $shouldRenderPastDueBanner, should render unpaid banner: $shouldRenderUnpaidBanner, should render database prompt banner: $shouldRenderDatabasePromptBanner",
```
nit: I might be nit-picking but I don't follow `shouldDoXXX` tests. I think what we actually want to test is different combinations of token features and the `isAdmin` flag, and that you get expected results there. Now the test looks like if `shouldRenderUnpaidBanner` is true, we render the unpaid banner. That's not what we really want to test. Should be fairly straightforward change if you remove the `should` flags and set the assigned properties directly.

ranquild replied:
Other tests are implemented in the way I'd expect, if we can fix this test suite it would be great

WiNloSt replied:
I was thinking about that too. But since there are tests that cover some combinations of those flags in `DatabasePromptBanner.unit.spec.tsx` and `PaymentBanner.unit.spec.tsx` it feels as if we're repeating a lot of that code here.

To be fair, I was also thinking about removing those 2 test suites and testing every combination here in `AppBanner.unit.spec.tsx` instead. Otherwise, we will just create a superset of tests from those 2 other test suites.

What do you think?

ranquild commented on frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:
```diff
@@ -0,0 +1,212 @@
+    | "Unpaid banner"
+    | "Database prompt banner",
+) {
+  if (renderingBanner === "Nothing") {
```
nit: You can get rid of this method if you split the `it.each` clause into 3 different clauses. Currently this is a bit hard to follow

WiNloSt replied:
I agree this is too hard to follow.

ranquild commented on frontend/src/metabase/nav/components/DatabasePromptBanner/DatabasePromptBanner.unit.spec.tsx:
```diff
@@ -0,0 +1,234 @@
+
+describe("DatabasePromptBanner", () => {
+  it("should not render for non-admin users without paid plan without connected databases", () => {
+    setup({
```
In contract to the previous comment, I really like these tests from the `setup` point of view - if you're an admin, on a paid plan, and have other databases and not just the sample one, you shouldn't see the paid banner. Makes sense from the product point of view.

ranquild commented on frontend/src/metabase/nav/components/PaymentBanner/PaymentBanner.unit.spec.tsx:
```diff
@@ -0,0 +1,126 @@
+  });
+
+  describe("non-admin users", () => {
+    it.each([
```
Also good tests

ranquild commented on frontend/src/metabase/nav/components/PaymentBanner/PaymentBanner.unit.spec.tsx:
```diff
@@ -0,0 +1,126 @@
+      {
+        isAdmin: false,
+        tokenStatusStatus: "something-else",
+        shouldRenderPaymentBanner: false,
```
I'd call it something different to avoid the collision, e.g. `hasBanner` or something like this

WiNloSt replied:
Fixed

ranquild commented on frontend/src/metabase/selectors/selectors.unit.spec.ts:
```diff
@@ -43,3 +43,35 @@ describe("getUpgradeUrl", () => {
   });
 });
+
+describe("getIsPaidPlan", () => {
```
good tests

ranquild reviewed (approved):
Looks good! Left some comments about one of the test suites, please take a look

uladzimirdev commented on frontend/src/metabase/components/AppBanner/AppBanner.tsx:
```diff
@@ -0,0 +1,28 @@
+    return (
+      <PaymentBanner isAdmin={isAdmin} tokenStatusStatus={tokenStatusStatus} />
+    );
+  } else {
```
you don't need `else` here

WiNloSt replied:
Ah yep that'll be cleaner 👍

WiNloSt replied:
fixed

cdeweyx reviewed (approved):
Works as expected, approved pending additional case handling
